File: lekcje/02_listy/przyklady/lista_c/lista_polaczona.c
#include "lista_polaczona.h"
#include <stdio.h>
#include <stdlib.h>

// Funkcja tworzaca nowy wezel z podana wartoscia.
Wezel *utworz_wezel(int wartosc) {
  Wezel *nowy_wezel = (Wezel *)malloc(sizeof(Wezel));
  if (nowy_wezel == NULL) {
    fprintf(stderr, "Nie udalo sie zaalokowac pamieci dla nowego wezla.\n");
    exit(EXIT_FAILURE);
  }
  nowy_wezel->wartosc = wartosc;
  nowy_wezel->nastepny = NULL;
  return nowy_wezel;
}

// Funkcja dodajaca nowy wezel na koncu listy.
void dodaj_na_koniec(Wezel **glowa, int wartosc) {
  Wezel *nowy_wezel = utworz_wezel(wartosc);
  if (*glowa == NULL) {
    *glowa = nowy_wezel;
    return;
  }
  Wezel *ostatni = *glowa;
  while (ostatni->nastepny != NULL) {
    ostatni = ostatni->nastepny;
  }
  ostatni->nastepny = nowy_wezel;
}

// Funkcja wypisujaca wszystkie elementy listy.
void wypisz_liste(Wezel *glowa) {
  Wezel *aktualny = glowa;
  while (aktualny != NULL) {
    printf("%d -> ", aktualny->wartosc);
    aktualny = aktualny->nastepny;
  }
  printf("NULL\n");
}

// Funkcja usuwajaca wszystkie elementy listy i zwalniajaca pamiec.
void usun_liste(Wezel **glowa) {
  Wezel *aktualny = *glowa;
  Wezel *nastepny;
  while (aktualny != NULL) {
    nastepny = aktualny->nastepny;
    free(aktualny);
    aktualny = nastepny;
  }
  *glowa = NULL;
}

File: lekcje/02_listy/przyklady/lista_python/lista_polaczona.py
# Klasa Wezel reprezentujaca pojedynczy element listy
class Wezel:
    def __init__(self, wartosc):
        self.wartosc = wartosc
        self.nastepny = None


# Klasa ListaPolaczona implementujaca liste jednokierunkowa
class ListaPolaczona:
    def __init__(self):
        self.glowa = None

    # Metoda dodajaca nowy wezel na koncu listy
    def dodaj_na_koniec(self, wartosc):
        nowy_wezel = Wezel(wartosc)
        if self.glowa is None:
            self.glowa = nowy_wezel
            return
        ostatni = self.glowa
        while ostatni.nastepny:
            ostatni = ostatni.nastepny
        ostatni.nastepny = nowy_wezel

    # Metoda wypisujaca wszystkie elementy listy
    def wypisz_liste(self):
        aktualny = self.glowa
        while aktualny:
            print(aktualny.wartosc, end=" -> ")
            aktualny = aktualny.nastepny
        print("None")

    # Metoda usuwajaca wszystkie elementy listy
    def usun_liste(self):
        self.glowa = None


# Funkcja main demonstrujaca uzycie listy
def main():
    lista = ListaPolaczona()
    lista.dodaj_na_koniec(1)
    lista.dodaj_na_koniec(2)
    lista.dodaj_na_koniec(3)

    print("Lista polaczona: ")
    lista.wypisz_liste()

    lista.usun_liste()


# Uruchomienie funkcji main
if __name__ == "__main__":
    main()
